README
The GNU Hurd
============

This is not the GNU Hurd.  But it might become it at some day.

Although this is called the Hurd on L4, L4 is only used as a hardware
abstraction: the kernel is Viengoos and no features of L4 are relied
on in a manner inconsistent with Viengoos' design.  If the design
proves tenable and is accepted by the community, the L4 dependency
will eventually be removed.


How to build
------------

You need at least automake 1.9 and autoconf 2.57.

To build the documentation, you also need pdflatex and some supporting
packages.  On Ubuntu Hardy (probably pretty much the same for Debian
and other Ubuntu systems), you'll need these packages installed for
the documentation build to succeed:

  * texlive-fonts-recommended
  * texlive-latex-base
  * texlive-latex-extra
  * texlive-latex-recommended
  * texlive-science


Building the code is straightforward:

$ autoreconf -i
$ mkdir test build
$ cd test/
$ ../configure --enable-l4 --host=i686-pc-viengoos-gnu \
	--enable-tests
$ make check
$ cd ../build/
$ ../configure --enable-l4 --host=i686-pc-viengoos-gnu \
	--with-newlib
$ make [to build the cross compiler]
$ make [to build Viengoos proper]
$ mkdir /l4/boot
$ install -s laden/laden /l4/boot
$ install -s viengoos/viengoos /l4/boot
$ install -s hieronymus/hieronymus /l4/boot

When configuring, you can use --enable-runtime-checks=N, 0 <= N <= 3,
to control which run-checks are executed.  Level 3 includes very
expensive consistency checks; level 0 disables all consistency checks,
the trace buffer and all debugging output except that at level 0.


If you happen to be on a multi-processor system, you can run ``make -j'' to
have several target be built in parallel.  This may save some time.  It might
even shorten the building time on one-processor system.  (No specific
measurements exist currently.)


Documentation
-------------

A ``Viengoos Developer Reference'' is available in the `doc' subdirectory.  A
PDF will be created during the build process if pdflatex is found.

Also, the header files are a good source of documentation.


L4
--

Viengoos uses Pistachio as a hardware abstraction.

Download Pistachio by cloning upstream's mercurial repository.  You
can find instructions on how to do this at:

	http://www.l4ka.org/projects/pistachio/download.php

Enter the kernel directory and run `make' to set up a build
environment.  Configure the source.  Select your platform.  Optionally
enabled the kernel debugger.  And, be sure to enable the following
option:

  Kernel -> Experimental Features -> Pager ExchangeRegisters

You'll also need to build sigma0.  It is in the user
directory.

sigma0 must be built with a different link base from the default
(0x20000), because this conflicts with GRUB.  I have successfully used
a link base of 0x40000.  Use the configure option
--with-s0-linkbase=0x40000 when configuring the user space of the L4
distribution.

You only need the ia32-kernel and sigma0 binary from the L4
distribution.  You may want to install the other programs and
utilities as well to test your L4 kernel, though.


Booting
-------

GRUB can boot Viengoos.  Set up GRUB to boot laden as the kernel, and
load the L4 kernel, sigma0, viengoos and hieronymus as modules (in
that order).

Try the debug option to laden (-D) and viengoos (-D 3) to see some
output from them.  They are relatively silent by default.

Here is an example menu.lst file for GNU GRUB:

title Viengoos
root (hd0,0)
kernel /boot/laden -D
module /boot/ia32-kernel
module /boot/sigma0
module /boot/viengoos -D 3
module /boot/hieronymus -D 3


Using the Cross-Compiler
------------------------

Does the package you want to cross-compile use the GNU build system
(`configure' script)?  If not, then...  If yes, then proceed as follows.  Do
something like ``PATH=$PATH:/where/the/viengoos/build/is/bin'' to make sure
that `i686-pc-viengoos-gnu-gcc' can be found via `$PATH'.  As the master GNU
system-guessing scripts don't know about Viengoos (yet), you have to link the
package's `config.guess' and `config.sub' to the Viengoos ones.  That's what
`fix_config_guess_and_sub' (see `Makefoo.am') does, used in
`newlib/Makefile.am', for example.  After that, configuring the to-be-built
package with ``--host=i686-pc-viengoos-gnu'' is all that *should* be needed,
for a portable package.  Otherwise you'll see things explode.




ChangeLog
2009-01-16  Neal H. Walfield  <neal@gnu.org>

	* Makefile.am (L4_SUBDIRS): New variable.  Set to l4 directories
	if USE_L4 is true, otherwise leave empty.
	(SUBDIRS): Remove libl4 and laden.  Replace with $(USE_L4).
	Rearrange order.

2009-01-16  Neal H. Walfield  <neal@gnu.org>

	* configure.ac (--enable-l4-abi): Replace option with...
	(--enable-l4): ... this option.
	(l4_abi): Set appropriately.
	(USE_L4): Define as a preprocessor macro (if appropriate) and an
	automake conditional.
	* README: Update build instructions.

2008-11-19  Neal H. Walfield  <neal@gnu.org>

	* configure.ac: Rename amd64 to x86_64.  Update users.

2008-12-17  Neal H. Walfield  <neal@gnu.org>

	* libviengoos: New directory.
	* Makefile.am (SUBDIRS): Add libviengoos.
	* configure.ac: Include libviengoos/headers.m4.  Generate
	libviengoos/Makefile.

2008-12-12  Neal H. Walfield  <neal@gnu.org>

	* README: Move build requirements prior to build instructions.
	Suggested by: Ben Asselstine <benasselstine@gmail.com>.

2008-11-20  Neal H. Walfield  <neal@gnu.org>

	* configure.ac: Don't look for inkscape.
	* README: Don't mention inkscape.

2008-12-11  Neal H. Walfield  <neal@gnu.org>

	* configure.ac (C_CHECKS): Add `-fstack-protector-all' for all
	runtime check levels except level 0.

2008-12-04  Neal H. Walfield  <neal@gnu.org>

	* configure.ac (USE_L4): AC_DEFINE it.
	(USER_CPPFLAGS): Add -DUSE_L4.
	(KERNEL_CPPFLAGS): Likewise.

2008-12-04  Neal H. Walfield  <neal@gnu.org>

	* Makefile.am: Don't build libstdc++.

2008-11-18  Neal H. Walfield  <neal@gnu.org>

	* README: Improve text.

2008-11-11  Neal H. Walfield  <neal@gnu.org>

	* README: Remove reference to patching Pistachio and libl4.

2008-11-11  Neal H. Walfield  <neal@gnu.org>

	* configure.ac: Generate benchmarks/sqlite/Makefile.

2008-11-11  Neal H. Walfield  <neal@gnu.org>

	* Makefile.am (libc-stmp): Only touch sysroot/lib/libc.a, don't
	overwrite it.

2008-11-03  Neal H. Walfield  <neal@gnu.org>

	* Revert last change.

2008-11-03  Neal H. Walfield  <neal@gnu.org>

	* Makefile.am (libc-stmp): Don't create an empty
	sysroot/lib/libc.a.

2008-08-22  Neal H. Walfield  <neal@gnu.org>

	* libhurd-cap: Remove.
	* libhurd-cap-server: Likewise.
	* physmem: Likewise.
	* task: Likewise.
	* wortel: Likewise.

2008-06-27  Neal H. Walfield  <neal@gnu.org>

	* configure.ac (--enable-runtime-checks): If zero, don't just
	define DEBUG_ELIDE, define it to 1.

2008-06-26  Neal H. Walfield  <neal@gnu.org>

	* configure.ac: Spell DEBUG_ELIDE correctly.

2008-06-22  Neal H. Walfield  <neal@gnu.org>

	* configure.ac (--enable-runtime-checks): Fix spelling.

2008-06-21  Thomas Schwinge  <tschwinge@gnu.org>

	* README (Using the Cross-Compiler): New section.

2008-06-20  Neal H. Walfield  <neal@gnu.org>

	* configure.ac (--enable-runtime-checks): New command line option.
	(USER_CPPFLAGS, KERNEL_CPPFLAGS): Add $CPP_CHECKS.
	(USER_CFLAGS, KERNEL_CFLAGS): Add $C_CHECKS.

2008-06-20  Thomas Schwinge  <tschwinge@gnu.org>

	* README (Documentation): New section.

2008-06-18  Thomas Schwinge  <tschwinge@gnu.org>

	* Makefile.am [HAVE_CROSS] (all-local): Depend on libstdc++.

2008-06-05  Neal H. Walfield  <neal@gnu.org>

	* libc.a.in: Remove -lhurd-cap.
	* Makefile.am (libc-stmp): Remove dependency on
	viengoos/libhurd-cap.a.

2008-06-01  Thomas Schwinge  <tschwinge@gnu.org>

	* Makefoo.am (PATH): Move definition...
	* configure.ac: ... here.

2008-05-31  Thomas Schwinge  <tschwinge@gnu.org>

	* configure.ac (USER_CFLAGS, KERNEL_CFLAGS, CHECK_CFLAGS): Don't
	specify ``-fno-stack-protector''.

	* configure.ac (ABS_BUILDDIR): Use a more portable way to set.

2008-05-30  Neal H. Walfield  <neal@gnu.org>

	* benchmarks: New directory.
	* configure.ac: Generate benchmarks/Makefile and
	benchmarks/boehm-gc/Makefile.
	* Makefile.am (SUBDIRS): Add benchmarks.

2008-05-30  Thomas Schwinge  <tschwinge@gnu.org>

	* configure.ac (missing_progs): Rename to...
	(missing_doc_progs): ... this.
	(DOC): New variable.  Used to indicate whether to documentation
	files can be built.
	Make it non-fatal if the documentation can't be built.
	* Makefile.am (SUBDIRS): Use $(DOC).

	* configure.ac (AM_MAINTAINER_MODE): Get rid of it.
	* README: Reflect that.

	* README: Suggest ``make -j''.

2008-05-30  Neal H. Walfield  <neal@gnu.org>

	* configure.ac: Call AM_PROG_CC_C_O.

2008-05-30  Neal H. Walfield  <neal@gnu.org>

	* Makefoo.am (abs_builddir): Define.

2008-05-30  Thomas Schwinge  <tschwinge@gnu.org>

	* configure.in (libc_cv_gnu89_inline): Fill depending on a compile-time
	test, as per glibc HEAD, 2007-11-07.
	(USER_CFLAGS, KERNEL_CFLAGS, CHECK_CFLAGS): Use it.
	
2008-05-29  Thomas Schwinge  <tschwinge@gnu.org>

	* configure.ac: Include `newlib/headers.m4' and
	`libc-parts/headers.m4'.
	(ccheaders, SYSINCLUDES): Remove.
	(USER_CPPFLAGS, USER_LDFLAGS, KERNEL_CPPFLAGS): Adapt to building with
	a cross compiler.
	(USER_LDADD): Remove.
	(CHECK_CPPFLAGS): Define `_GNU_SOURCE' and `_ENABLE_TESTS'.
	(CHECK_CFLAGS): New.
	(HAVE_CROSS, TRAVERSE_HIERARCHY): New conditionals.
	(--enable-tests, ENABLE_TESTS): New command line option with
	conditional.
	(libc.a.in): Instantiate as `sysroot/lib/libc.a'.
	* libc.a.in: Rewritten to no longer specify file paths, but instead
	linker switches.
	* Makefile.am (SUBDIRS): Only recurse into if [TRAVERSE_HIERARCHY].
	(libc-stmp): Only consider if [HAVE_CROSS].
	[! HAVE_CROSS && ! ENABLE_TESTS]: Build `cross-compiler'.
	(cross-compiler): New target.
	Include `Makefoo.am', `binutils/Makefrag.am' and `gcc/Makefrag.am'.
	* Makefoo.am: New file.
	* binutils: New directory.
	* gcc: Likewise.
	* README: Update.

	* build-aux/config.sub: Detect *-*-viengoos-gnu systems.

	* build-aux/config.guess: New file, from today's config git trunk.
	* build-aux/config.sub: Likewise.

	* configure.ac (AC_CONFIG_AUX_DIR): Instantiate.
	(AM_INIT_AUTOMAKE): Add a number of options.
	* README: Adapt.

2008-05-27  Neal H. Walfield  <neal@gnu.org>


hurd/math.h
/* l4/math.h - Public interface to Hurd mathematical support functions.
   Copyright (C) 2003, 2004, 2007, 2008 Free Software Foundation, Inc.
   Written by Marcus Brinkmann <marcus@gnu.org>.

   This file is part of the GNU Hurd.
 
   The GNU Hurd is free software; you can redistribute it and/or
   modify it under the terms of the GNU Lesser General Public License
   as published by the Free Software Foundation; either version 2.1 of
   the License, or (at your option) any later version.
 
   The GNU Hurd is distributed in the hope that it will be
   useful, but WITHOUT ANY WARRANTY; without even the implied warranty
   of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
   GNU Lesser General Public License for more details.
 
   You should have received a copy of the GNU Lesser General Public
   License along with the GNU L4 library; if not, write to the Free
   Software Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA
   02111-1307, USA.  */

#ifndef _HURD_MATH_H
#define _HURD_MATH_H

#include <stdint.h>
#include <bits/wordsize.h>
#include <hurd/bits/math.h>

/* Return 0 if DATA is 0, or the bit number of the most significant
   bit set in DATA.  The least significant bit is 1, the most
   significant bit 32 resp. 64.  */
static inline uintptr_t
vg_msb (uintptr_t data)
{
  if (__builtin_constant_p (data))
    {
#define __VG_MSB_TRY(b) else if (data < (1 << (b))) return (b)
#define __VG_MSB_IS(b) else return (b)

      if (!data)
	return 0;
      __VG_MSB_TRY(1); __VG_MSB_TRY(2); __VG_MSB_TRY(3); __VG_MSB_TRY(4);
      __VG_MSB_TRY(5); __VG_MSB_TRY(6); __VG_MSB_TRY(7); __VG_MSB_TRY(8);
      __VG_MSB_TRY(9); __VG_MSB_TRY(10); __VG_MSB_TRY(11); __VG_MSB_TRY(12);
      __VG_MSB_TRY(13); __VG_MSB_TRY(14); __VG_MSB_TRY(15); __VG_MSB_TRY(16);
      __VG_MSB_TRY(17); __VG_MSB_TRY(18); __VG_MSB_TRY(19); __VG_MSB_TRY(20);
      __VG_MSB_TRY(21); __VG_MSB_TRY(22); __VG_MSB_TRY(23); __VG_MSB_TRY(24);
      __VG_MSB_TRY(25); __VG_MSB_TRY(26); __VG_MSB_TRY(27); __VG_MSB_TRY(28);
      __VG_MSB_TRY(29); __VG_MSB_TRY(30); __VG_MSB_TRY(31);
#if __WORDSIZE == 32
      __VG_MSB_IS(32);
#else
      __VG_MSB_TRY(32); __VG_MSB_TRY(33); __VG_MSB_TRY(34); __VG_MSB_TRY(35);
      __VG_MSB_TRY(36); __VG_MSB_TRY(37); __VG_MSB_TRY(38); __VG_MSB_TRY(39);
      __VG_MSB_TRY(40); __VG_MSB_TRY(41); __VG_MSB_TRY(42); __VG_MSB_TRY(43);
      __VG_MSB_TRY(44); __VG_MSB_TRY(45); __VG_MSB_TRY(46); __VG_MSB_TRY(47);
      __VG_MSB_TRY(48); __VG_MSB_TRY(49); __VG_MSB_TRY(50); __VG_MSB_TRY(51);
      __VG_MSB_TRY(52); __VG_MSB_TRY(53); __VG_MSB_TRY(54); __VG_MSB_TRY(55);
      __VG_MSB_TRY(56); __VG_MSB_TRY(57); __VG_MSB_TRY(58); __VG_MSB_TRY(59);
      __VG_MSB_TRY(60); __VG_MSB_TRY(61); __VG_MSB_TRY(62); __VG_MSB_TRY(63);
      __VG_MSB_IS(64);
#endif
    }

  if (__builtin_expect (data != 0, 1))
    return __VG_msb (data);
  else
    return 0;
}

static inline uintptr_t
vg_msb64 (uint64_t data)
{
#if __WORDSIZE == 64
  return vg_msb (data);
#else
  int d = vg_msb (data >> 32);
  if (d)
    return d + 32;
  return vg_msb (data);
#endif
}

/* Return 0 if DATA is 0, or the bit number of the least significant
   bit set in DATA.  The least significant bit is 1, the most
   significant bit 32 resp. 64.  */
static inline uintptr_t
vg_lsb (uintptr_t data)
{
  if (__builtin_constant_p (data))
    {
#define __VG_LSB_TRY(b) else if (data >= (1 << (b - 1))) return (b)
#define __VG_LSB_IS(b) else return (b)

      if (!data)
	return 0;
#if __WORDSIZE == 64
      __VG_LSB_TRY(64); __VG_LSB_TRY(63); __VG_LSB_TRY(62); __VG_LSB_TRY(61);
      __VG_LSB_TRY(60); __VG_LSB_TRY(59); __VG_LSB_TRY(58); __VG_LSB_TRY(57);
      __VG_LSB_TRY(56); __VG_LSB_TRY(55); __VG_LSB_TRY(54); __VG_LSB_TRY(53);
      __VG_LSB_TRY(52); __VG_LSB_TRY(51); __VG_LSB_TRY(50); __VG_LSB_TRY(49);
      __VG_LSB_TRY(48); __VG_LSB_TRY(47); __VG_LSB_TRY(46); __VG_LSB_TRY(45);
      __VG_LSB_TRY(44); __VG_LSB_TRY(43); __VG_LSB_TRY(42); __VG_LSB_TRY(41);
      __VG_LSB_TRY(40); __VG_LSB_TRY(39); __VG_LSB_TRY(38); __VG_LSB_TRY(37);
      __VG_LSB_TRY(36); __VG_LSB_TRY(35); __VG_LSB_TRY(34); __VG_LSB_TRY(33);
#endif
      __VG_LSB_TRY(32); __VG_LSB_TRY(31); __VG_LSB_TRY(30); __VG_LSB_TRY(29);
      __VG_LSB_TRY(28); __VG_LSB_TRY(27); __VG_LSB_TRY(26); __VG_LSB_TRY(25);
      __VG_LSB_TRY(24); __VG_LSB_TRY(23); __VG_LSB_TRY(22); __VG_LSB_TRY(21);
      __VG_LSB_TRY(20); __VG_LSB_TRY(19); __VG_LSB_TRY(18); __VG_LSB_TRY(17);
      __VG_LSB_TRY(16); __VG_LSB_TRY(15); __VG_LSB_TRY(14); __VG_LSB_TRY(13);
      __VG_LSB_TRY(12); __VG_LSB_TRY(11); __VG_LSB_TRY(10); __VG_LSB_TRY(9);
      __VG_LSB_TRY(8); __VG_LSB_TRY(7); __VG_LSB_TRY(6); __VG_LSB_TRY(5);
      __VG_LSB_TRY(4); __VG_LSB_TRY(3); __VG_LSB_TRY(2); __VG_LSB_IS(1);
    }

  if (__builtin_expect (data != 0, 1))
    return __VG_lsb (data);
  else
    return 0;
}

static inline uintptr_t
vg_lsb64 (uint64_t data)
{
#if __WORDSIZE == 64
  return vg_lsb (data);
#else
  int d = vg_lsb (data);
  if (d)
    return d;
  d = vg_lsb (data >> 32);
  if (d)
    return d + 32;
  return 0;
#endif
}

#endif
